db/document.go

```diff
@@ -106,6 +109,86 @@ func UnmarshalDocumentSyncData(data []byte, needHistory bool) (*syncData, error)
 	return root.SyncData, nil
 }
 
+// Unmarshals sync metadata for a document arriving via DCP.  Includes handling for xattr content
+// being included in data.  If not present in either xattr or document body, returns nil but no error.
+// We may need to
+func UnmarshalDocumentSyncDataFromFeed(data []byte, dataType uint8, needHistory bool) (result *syncData, err error) {
+
+	var body []byte
+
+	// If attr datatype flag is set, data includes both xattrs and document body.  Check for presence of sync xattr.
+	// Note that there could be a non-sync xattr present
+	if dataType&base.MemcachedDataTypeXattr != 0 {
+		var syncXattr []byte
+		body, syncXattr, err = parseXattrStreamData(KSyncXattrName, data)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the sync xattr is present, use that to build syncData
+		if syncXattr != nil && len(syncXattr) > 0 {
+			result = &syncData{}
+			if needHistory {
+				result.History = make(RevTree)
+			}
+			err = json.Unmarshal(syncXattr, result)
+			if err != nil {
+				return nil, err
+			}
+			return result, nil
+		}
+	} else {
+		// Xattr flag not set - data is just the document body
+		body = data
+	}
+
+	// Non-xattr data, or sync xattr not present.  Attempt to retrieve sync metadata from document body
+	return UnmarshalDocumentSyncData(body, needHistory)
+
+}
+
+// parseXattrStreamData returns the raw bytes of the body and the requested xattr (when present) from the raw DCP data bytes.
+// Details on format: https://docs.google.com/document/d/18UVa5j8KyufnLLy29VObbWRtoBn9vs8pcxttuMt6rz8/edit#heading=h.caqiui1pmmmb
+func parseXattrStreamData(xattrName string, data []byte) (body []byte, xattr []byte, err error) {
+
+	xattrsLen := binary.BigEndian.Uint32(data[0:4])
```

Comment:
With the additional documentation copied into the code, I think it's clearer what xattrsLen refers to (all the xattrs).  

I don't think the '4' should be a constant - I think it would actually make harder to work out what the code is doing (having to go look up what the constant is set to).